File: src/daft-sql/src/modules/numeric.rs
use daft_dsl::{ExprRef, LiteralValue};
use daft_functions::numeric::{
    abs::abs,
    ceil::ceil,
    exp::exp,
    floor::floor,
    log::{ln, log, log10, log2},
    round::round,
    sign::sign,
    sqrt::sqrt,
    trigonometry::{
        arccos, arccosh, arcsin, arcsinh, arctan, arctanh, atan2, cos, cot, degrees, radians, sin,
        tan,
    },
};

use super::SQLModule;
use crate::{
    ensure,
    error::{PlannerError, SQLPlannerResult},
    functions::{SQLFunction, SQLFunctions},
    invalid_operation_err,
};

pub struct SQLModuleNumeric;

impl SQLModule for SQLModuleNumeric {
    fn register(parent: &mut SQLFunctions) {
        parent.add_fn("abs", SQLNumericExpr::Abs);
        parent.add_fn("ceil", SQLNumericExpr::Ceil);
        parent.add_fn("floor", SQLNumericExpr::Floor);
        parent.add_fn("sign", SQLNumericExpr::Sign);
        parent.add_fn("round", SQLNumericExpr::Round);
        parent.add_fn("sqrt", SQLNumericExpr::Sqrt);
        parent.add_fn("sin", SQLNumericExpr::Sin);
        parent.add_fn("cos", SQLNumericExpr::Cos);
        parent.add_fn("tan", SQLNumericExpr::Tan);
        parent.add_fn("cot", SQLNumericExpr::Cot);
        parent.add_fn("asin", SQLNumericExpr::ArcSin);
        parent.add_fn("acos", SQLNumericExpr::ArcCos);
        parent.add_fn("atan", SQLNumericExpr::ArcTan);
        parent.add_fn("atan2", SQLNumericExpr::ArcTan2);
        parent.add_fn("radians", SQLNumericExpr::Radians);
        parent.add_fn("degrees", SQLNumericExpr::Degrees);
        parent.add_fn("log2", SQLNumericExpr::Log2);
        parent.add_fn("log10", SQLNumericExpr::Log10);
        parent.add_fn("log", SQLNumericExpr::Log);
        parent.add_fn("ln", SQLNumericExpr::Ln);
        parent.add_fn("exp", SQLNumericExpr::Exp);
        parent.add_fn("atanh", SQLNumericExpr::ArcTanh);
        parent.add_fn("acosh", SQLNumericExpr::ArcCosh);
        parent.add_fn("asinh", SQLNumericExpr::ArcSinh);
    }
}
enum SQLNumericExpr {
    Abs,
    Ceil,
    Exp,
    Floor,
    Round,
    Sign,
    Sqrt,
    Sin,
    Cos,
    Tan,
    Cot,
    ArcSin,
    ArcCos,
    ArcTan,
    ArcTan2,
    Radians,
    Degrees,
    Log,
    Log2,
    Log10,
    Ln,
    ArcTanh,
    ArcCosh,
    ArcSinh,
}

impl SQLFunction for SQLNumericExpr {
    fn to_expr(
        &self,
        inputs: &[sqlparser::ast::FunctionArg],
        planner: &crate::planner::SQLPlanner,
    ) -> SQLPlannerResult<ExprRef> {
        let inputs = self.args_to_expr_unnamed(inputs, planner)?;
        to_expr(self, inputs.as_slice())
    }

    fn docstrings(&self, _alias: &str) -> String {
        let docstring = match self {
            Self::Abs => "Gets the absolute value of a number.",
            Self::Ceil => "Rounds a number up to the nearest integer.",
            Self::Exp => "Calculates the exponential of a number (e^x).",
            Self::Floor => "Rounds a number down to the nearest integer.",
            Self::Round => "Rounds a number to a specified number of decimal places.",
            Self::Sign => "Returns the sign of a number (-1, 0, or 1).",
            Self::Sqrt => "Calculates the square root of a number.",
            Self::Sin => "Calculates the sine of an angle in radians.",
            Self::Cos => "Calculates the cosine of an angle in radians.",
            Self::Tan => "Calculates the tangent of an angle in radians.",
            Self::Cot => "Calculates the cotangent of an angle in radians.",
            Self::ArcSin => "Calculates the inverse sine (arc sine) of a number.",
            Self::ArcCos => "Calculates the inverse cosine (arc cosine) of a number.",
            Self::ArcTan => "Calculates the inverse tangent (arc tangent) of a number.",
            Self::ArcTan2 => {
                "Calculates the angle between the positive x-axis and the ray from (0,0) to (x,y)."
            }
            Self::Radians => "Converts an angle from degrees to radians.",
            Self::Degrees => "Converts an angle from radians to degrees.",
            Self::Log => "Calculates the natural logarithm of a number.",
            Self::Log2 => "Calculates the base-2 logarithm of a number.",
            Self::Log10 => "Calculates the base-10 logarithm of a number.",
            Self::Ln => "Calculates the natural logarithm of a number.",
            Self::ArcTanh => "Calculates the inverse hyperbolic tangent of a number.",
            Self::ArcCosh => "Calculates the inverse hyperbolic cosine of a number.",
            Self::ArcSinh => "Calculates the inverse hyperbolic sine of a number.",
        };
        docstring.to_string()
    }

    fn arg_names(&self) -> &'static [&'static str] {
        match self {
            Self::Abs
            | Self::Ceil
            | Self::Floor
            | Self::Sign
            | Self::Sqrt
            | Self::Sin
            | Self::Cos
            | Self::Tan
            | Self::Cot
            | Self::ArcSin
            | Self::ArcCos
            | Self::ArcTan
            | Self::Radians
            | Self::Degrees
            | Self::Log2
            | Self::Log10
            | Self::Ln
            | Self::ArcTanh
            | Self::ArcCosh
            | Self::ArcSinh => &["input"],
            Self::Log => &["input", "base"],
            Self::Round => &["input", "precision"],
            Self::Exp => &["input", "exponent"],
            Self::ArcTan2 => &["y", "x"],
        }
    }
}

fn to_expr(expr: &SQLNumericExpr, args: &[ExprRef]) -> SQLPlannerResult<ExprRef> {
    match expr {
        SQLNumericExpr::Abs => {
            ensure!(args.len() == 1, "abs takes exactly one argument");
            Ok(abs(args[0].clone()))
        }
        SQLNumericExpr::Ceil => {
            ensure!(args.len() == 1, "ceil takes exactly one argument");
            Ok(ceil(args[0].clone()))
        }
        SQLNumericExpr::Floor => {
            ensure!(args.len() == 1, "floor takes exactly one argument");
            Ok(floor(args[0].clone()))
        }
        SQLNumericExpr::Sign => {
            ensure!(args.len() == 1, "sign takes exactly one argument");
            Ok(sign(args[0].clone()))
        }
        SQLNumericExpr::Round => {
            ensure!(args.len() == 2, "round takes exactly two arguments");
            let precision = match args[1].as_ref().as_literal() {
                Some(LiteralValue::Int32(i)) => *i,
                Some(LiteralValue::UInt32(u)) => *u as i32,
                Some(LiteralValue::Int64(i)) => *i as i32,
                _ => invalid_operation_err!("round precision must be an integer"),
            };
            Ok(round(args[0].clone(), precision))
        }
        SQLNumericExpr::Sqrt => {
            ensure!(args.len() == 1, "sqrt takes exactly one argument");
            Ok(sqrt(args[0].clone()))
        }
        SQLNumericExpr::Sin => {
            ensure!(args.len() == 1, "sin takes exactly one argument");
            Ok(sin(args[0].clone()))
        }
        SQLNumericExpr::Cos => {
            ensure!(args.len() == 1, "cos takes exactly one argument");
            Ok(cos(args[0].clone()))
        }
        SQLNumericExpr::Tan => {
            ensure!(args.len() == 1, "tan takes exactly one argument");
            Ok(tan(args[0].clone()))
        }
        SQLNumericExpr::Cot => {
            ensure!(args.len() == 1, "cot takes exactly one argument");
            Ok(cot(args[0].clone()))
        }
        SQLNumericExpr::ArcSin => {
            ensure!(args.len() == 1, "asin takes exactly one argument");
            Ok(arcsin(args[0].clone()))
        }
        SQLNumericExpr::ArcCos => {
            ensure!(args.len() == 1, "acos takes exactly one argument");
            Ok(arccos(args[0].clone()))
        }
        SQLNumericExpr::ArcTan => {
            ensure!(args.len() == 1, "atan takes exactly one argument");
            Ok(arctan(args[0].clone()))
        }
        SQLNumericExpr::ArcTan2 => {
            ensure!(args.len() == 2, "atan2 takes exactly two arguments");
            Ok(atan2(args[0].clone(), args[1].clone()))
        }
        SQLNumericExpr::Degrees => {
            ensure!(args.len() == 1, "degrees takes exactly one argument");
            Ok(degrees(args[0].clone()))
        }
        SQLNumericExpr::Radians => {
            ensure!(args.len() == 1, "radians takes exactly one argument");
            Ok(radians(args[0].clone()))
        }
        SQLNumericExpr::Log2 => {
            ensure!(args.len() == 1, "log2 takes exactly one argument");
            Ok(log2(args[0].clone()))
        }
        SQLNumericExpr::Log10 => {
            ensure!(args.len() == 1, "log10 takes exactly one argument");
            Ok(log10(args[0].clone()))
        }
        SQLNumericExpr::Ln => {
            ensure!(args.len() == 1, "ln takes exactly one argument");
            Ok(ln(args[0].clone()))
        }
        SQLNumericExpr::Log => {
            ensure!(args.len() == 2, "log takes exactly two arguments");
            let base = args[1]
                .as_literal()
                .and_then(|lit| match lit {
                    LiteralValue::Float64(f) => Some(*f),
                    LiteralValue::Int32(i) => Some(f64::from(*i)),
                    LiteralValue::UInt32(u) => Some(f64::from(*u)),
                    LiteralValue::Int64(i) => Some(*i as f64),
                    LiteralValue::UInt64(u) => Some(*u as f64),
                    _ => None,
                })
                .ok_or_else(|| PlannerError::InvalidOperation {
                    message: "log base must be a float or a number".to_string(),
                })?;

            Ok(log(args[0].clone(), base))
        }
        SQLNumericExpr::Exp => {
            ensure!(args.len() == 1, "exp takes exactly one argument");
            Ok(exp(args[0].clone()))
        }
        SQLNumericExpr::ArcTanh => {
            ensure!(args.len() == 1, "atanh takes exactly one argument");
            Ok(arctanh(args[0].clone()))
        }
        SQLNumericExpr::ArcCosh => {
            ensure!(args.len() == 1, "acosh takes exactly one argument");
            Ok(arccosh(args[0].clone()))
        }
        SQLNumericExpr::ArcSinh => {
            ensure!(args.len() == 1, "asinh takes exactly one argument");
            Ok(arcsinh(args[0].clone()))
        }
    }
}

File: daft/functions/numeric.py
"""Numeric Functions."""

from __future__ import annotations

import math

from daft.expressions import Expression


def abs(expr: Expression) -> Expression:
    """Absolute of a numeric expression."""
    return Expression._call_builtin_scalar_fn("abs", expr)


def e() -> Expression:
    """Returns Euler's number (e = 2.71828...)."""
    return Expression._call_builtin_scalar_fn("e")


def ceil(expr: Expression) -> Expression:
    """The ceiling of a numeric expression."""
    return Expression._call_builtin_scalar_fn("ceil", expr)


def floor(expr: Expression) -> Expression:
    """The floor of a numeric expression."""
    return Expression._call_builtin_scalar_fn("floor", expr)


def clip(
    expr: Expression,
    min: Expression | None = None,
    max: Expression | None = None,
) -> Expression:
    """Clips an expression to the given minimum and maximum values.

    Args:
        expr: The expression to clip
        min: Minimum value to clip to. If None (or column value is Null), no lower clipping is applied.
        max: Maximum value to clip to. If None (or column value is Null), no upper clipping is applied.
    """
    return Expression._call_builtin_scalar_fn("clip", expr, min, max)


def sign(expr: Expression) -> Expression:
    """The sign of a numeric expression."""
    return Expression._call_builtin_scalar_fn("sign", expr)


def negate(expr: Expression) -> Expression:
    """The negative of a numeric expression."""
    return Expression._call_builtin_scalar_fn("negate", expr)


def round(expr: Expression, decimals: Expression | int = 0) -> Expression:
    """The round of a numeric expression.

    Args:
        expr: The expression to round
        decimals: number of decimal places to round to. Defaults to 0.
    """
    return Expression._call_builtin_scalar_fn("round", expr, decimals)


def sqrt(expr: Expression) -> Expression:
    """The square root of a numeric expression."""
    return Expression._call_builtin_scalar_fn("sqrt", expr)


def cbrt(expr: Expression) -> Expression:
    """The cube root of a numeric expression."""
    return Expression._call_builtin_scalar_fn("cbrt", expr)


def sin(expr: Expression) -> Expression:
    """The elementwise sine of a numeric expression."""
    return Expression._call_builtin_scalar_fn("sin", expr)


def cos(expr: Expression) -> Expression:
    """The elementwise cosine of a numeric expression."""
    return Expression._call_builtin_scalar_fn("cos", expr)


def tan(expr: Expression) -> Expression:
    """The elementwise tangent of a numeric expression."""
    return Expression._call_builtin_scalar_fn("tan", expr)


def csc(expr: Expression) -> Expression:
    """The elementwise cosecant of a numeric expression."""
    return Expression._call_builtin_scalar_fn("csc", expr)


def sec(expr: Expression) -> Expression:
    """The elementwise secant of a numeric expression."""
    return Expression._call_builtin_scalar_fn("sec", expr)


def cot(expr: Expression) -> Expression:
    """The elementwise cotangent of a numeric expression."""
    return Expression._call_builtin_scalar_fn("cot", expr)


def sinh(expr: Expression) -> Expression:
    """The elementwise hyperbolic sine of a numeric expression."""
    return Expression._call_builtin_scalar_fn("sinh", expr)


def cosh(expr: Expression) -> Expression:
    """The elementwise hyperbolic cosine of a numeric expression."""
    return Expression._call_builtin_scalar_fn("cosh", expr)


def tanh(expr: Expression) -> Expression:
    """The elementwise hyperbolic tangent of a numeric expression."""
    return Expression._call_builtin_scalar_fn("tanh", expr)


def arcsin(expr: Expression) -> Expression:
    """The elementwise arc sine of a numeric expression."""
    return Expression._call_builtin_scalar_fn("arcsin", expr)


def arccos(expr: Expression) -> Expression:
    """The elementwise arc cosine of a numeric expression."""
    return Expression._call_builtin_scalar_fn("arccos", expr)


def arctan(expr: Expression) -> Expression:
    """The elementwise arc tangent of a numeric expression."""
    return Expression._call_builtin_scalar_fn("arctan", expr)


def arctan2(y: Expression, x: Expression) -> Expression:
    """Calculates the four quadrant arctangent of coordinates (y, x), in radians.

    * ``x = 0``, ``y = 0``: ``0``
    * ``x >= 0``: ``[-pi/2, pi/2]``
    * ``y >= 0``: ``(pi/2, pi]``
    * ``y < 0``: ``(-pi, -pi/2)``
    """
    return Expression._call_builtin_scalar_fn("arctan2", y, x)


def arctanh(expr: Expression) -> Expression:
    """The elementwise inverse hyperbolic tangent of a numeric expression."""
    return Expression._call_builtin_scalar_fn("arctanh", expr)


def arccosh(expr: Expression) -> Expression:
    """The elementwise inverse hyperbolic cosine of a numeric expression."""
    return Expression._call_builtin_scalar_fn("arccosh", expr)


def arcsinh(expr: Expression) -> Expression:
    """The elementwise inverse hyperbolic sine of a numeric expression."""
    return Expression._call_builtin_scalar_fn("arcsinh", expr)


def radians(expr: Expression) -> Expression:
    """The elementwise radians of a numeric expression."""
    return Expression._call_builtin_scalar_fn("radians", expr)


def degrees(expr: Expression) -> Expression:
    """The elementwise degrees of a numeric expression."""
    return Expression._call_builtin_scalar_fn("degrees", expr)


def log2(expr: Expression) -> Expression:
    """The elementwise log base 2 of a numeric expression."""
    return Expression._call_builtin_scalar_fn("log2", expr)


def log10(expr: Expression) -> Expression:
    """The elementwise log base 10 of a numeric expression."""
    return Expression._call_builtin_scalar_fn("log10", expr)


def log(expr: Expression, base: int | float = math.e) -> Expression:
    """The elementwise log with given base, of a numeric expression.

    Args:
        expr: The expression to take the logarithm of
        base: The base of the logarithm. Defaults to e.
    """
    return Expression._call_builtin_scalar_fn("log", expr, base)


def ln(expr: Expression) -> Expression:
    """The elementwise natural log of a numeric expression."""
    return Expression._call_builtin_scalar_fn("ln", expr)


def log1p(expr: Expression) -> Expression:
    """The ln(expr + 1) of a numeric expression."""
    return Expression._call_builtin_scalar_fn("log1p", expr)


def factorial(expr: Expression) -> Expression:
    """Returns the factorial of a non-negative integer."""
    return Expression._call_builtin_scalar_fn("factorial", expr)


def hypot(a: Expression, b: Expression) -> Expression:
    """Returns sqrt(a^2 + b^2), the Euclidean norm."""
    return Expression._call_builtin_scalar_fn("hypot", a, b)


def pi() -> Expression:
    """Returns the mathematical constant pi (3.14159...)."""
    return Expression._call_builtin_scalar_fn("pi")


def pow(base: Expression, expr: Expression) -> Expression:
    """The base^expr of a numeric expression."""
    return Expression._call_builtin_scalar_fn("pow", base, expr)


def power(base: Expression, expr: Expression) -> Expression:
    """The base^expr of a numeric expression."""
    return Expression._call_builtin_scalar_fn("power", base, expr)


def pmod(a: Expression, b: Expression) -> Expression:
    """Returns the positive modulo of ``a`` by ``b``.

    Computes ``r = a % b``; returns ``r`` when ``r >= 0`` and ``(r + b) % b`` otherwise.
    Examples: ``pmod(-7, 3) == 2``, ``pmod(7, -3) == 1``, ``pmod(-7, -3) == -1``.
    Returns NULL when ``b`` is 0.
    """
    return Expression._call_builtin_scalar_fn("pmod", a, b)


def exp(expr: Expression) -> Expression:
    """The e^expr of a numeric expression."""
    return Expression._call_builtin_scalar_fn("exp", expr)


def expm1(expr: Expression) -> Expression:
    """The e^expr - 1 of a numeric expression."""
    return Expression._call_builtin_scalar_fn("expm1", expr)


def between(expr: Expression, lower: Expression | int | float, upper: Expression | int | float) -> Expression:
    """Checks if values in the Expression are between lower and upper, inclusive.

    Args:
        expr: The expression to check
        lower: Lower bound (inclusive)
        upper: Upper bound (inclusive)

    Returns:
        Expression: Boolean Expression indicating whether values are between lower and upper, inclusive.

    Examples:
        >>> import daft
        >>> from daft.functions import between
        >>> df = daft.from_pydict({"data": [1, 2, 3, 4]})
        >>> df = df.select(between(df["data"], 1, 2))
        >>> df.collect()
        ╭───────╮
        │ data  │
        │ ---   │
        │ Bool  │
        ╞═══════╡
        │ true  │
        ├╌╌╌╌╌╌╌┤
        │ true  │
        ├╌╌╌╌╌╌╌┤
        │ false │
        ├╌╌╌╌╌╌╌┤
        │ false │
        ╰───────╯
        <BLANKLINE>
        (Showing first 4 of 4 rows)

    """
    expr = Expression._to_expression(expr)
    lower = Expression._to_expression(lower)
    upper = Expression._to_expression(upper)

    return Expression._from_pyexpr(expr._expr.between(lower._expr, upper._expr))


def bin(expr: Expression) -> Expression:
    """Returns the string representation of the binary value of an integer.

    Inputs are promoted to 64-bit before conversion; negatives produce
    64-character two's-complement strings (e.g. ``bin(-1)`` returns 64 ones).
    """
    return Expression._call_builtin_scalar_fn("bin", expr)


def conv(expr: Expression, from_base: int, to_base: int) -> Expression:
    """Converts a number from base ``from_base`` to base ``to_base`` (bases 2-36).

    Positive ``to_base`` interprets negative inputs as 64-bit two's complement
    (``conv("-1", 10, 16) == "FFFFFFFFFFFFFFFF"``); negative ``to_base`` returns
    a signed result (``conv("-1", 10, -16) == "-1"``). Trailing invalid characters
    are silently truncated (``conv("11abc", 10, 16) == "B"``). Returns NULL on
    out-of-range bases, on u64 overflow during parsing, or when a negated
    magnitude exceeds 2^63.
    """
    return Expression._call_builtin_scalar_fn("conv", expr, from_base, to_base)


def bround(expr: Expression, decimals: Expression | int = 0) -> Expression:
    """Rounds a numeric expression to ``decimals`` places using HALF_EVEN (banker's) rounding.

    Negative ``decimals`` rounds to powers of 10 above the decimal point
    (e.g. ``bround(125, -1) == 120`` because 12.5 rounds to even -> 12).

    Args:
        expr: The expression to round.
        decimals: Number of decimal places to round to. Defaults to 0.
    """
    return Expression._call_builtin_scalar_fn("bround", expr, decimals)


def hex(expr: Expression) -> Expression:
    """Converts an integer/string/binary expression to its uppercase hexadecimal string.

    For integer inputs, negatives are encoded as 64-bit two's complement
    (``hex(-1) == 'FFFFFFFFFFFFFFFF'``). For string and binary inputs, returns
    the uppercase hex of the underlying bytes (``hex('Spark') == '537061726B'``).
    """
    return Expression._call_builtin_scalar_fn("hex", expr)


def unhex(expr: Expression) -> Expression:
    r"""Inverse of :func:`hex`: decodes a hexadecimal string into binary bytes.

    Odd-length inputs are left-padded with ``'0'`` (``unhex('F') == b'\x0f'``).
    Returns NULL when the input contains characters outside ``[0-9a-fA-F]``.
    """
    return Expression._call_builtin_scalar_fn("unhex", expr)


def greatest(*exprs: Expression) -> Expression:
    """Returns the largest value among the inputs, skipping NULLs row-wise.

    Returns NULL only when all inputs in a row are NULL. Inputs are promoted
    to a common supertype before comparison. Requires at least one argument.

    Examples:
        >>> import daft
        >>> from daft.functions import greatest
        >>> df = daft.from_pydict({"a": [1, None, 3], "b": [2, 5, 1], "c": [None, 4, 6]})
        >>> df = df.with_column("g", greatest(df["a"], df["b"], df["c"]))
        >>> df.show()
        ╭───────┬───────┬───────┬───────╮
        │ a     ┆ b     ┆ c     ┆ g     │
        │ ---   ┆ ---   ┆ ---   ┆ ---   │
        │ Int64 ┆ Int64 ┆ Int64 ┆ Int64 │
        ╞═══════╪═══════╪═══════╪═══════╡
        │ 1     ┆ 2     ┆ None  ┆ 2     │
        ├╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┤
        │ None  ┆ 5     ┆ 4     ┆ 5     │
        ├╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┤
        │ 3     ┆ 1     ┆ 6     ┆ 6     │
        ╰───────┴───────┴───────┴───────╯
        <BLANKLINE>
        (Showing first 3 of 3 rows)
    """
    return Expression._call_builtin_scalar_fn("greatest", *exprs)


def least(*exprs: Expression) -> Expression:
    """Returns the smallest value among the inputs, skipping NULLs row-wise.

    Returns NULL only when all inputs in a row are NULL. Inputs are promoted
    to a common supertype before comparison. Requires at least one argument.

    Examples:
        >>> import daft
        >>> from daft.functions import least
        >>> df = daft.from_pydict({"a": [1, None, 3], "b": [2, 5, 1], "c": [None, 4, 6]})
        >>> df = df.with_column("l", least(df["a"], df["b"], df["c"]))
        >>> df.show()
        ╭───────┬───────┬───────┬───────╮
        │ a     ┆ b     ┆ c     ┆ l     │
        │ ---   ┆ ---   ┆ ---   ┆ ---   │
        │ Int64 ┆ Int64 ┆ Int64 ┆ Int64 │
        ╞═══════╪═══════╪═══════╪═══════╡
        │ 1     ┆ 2     ┆ None  ┆ 1     │
        ├╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┤
        │ None  ┆ 5     ┆ 4     ┆ 4     │
        ├╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┤
        │ 3     ┆ 1     ┆ 6     ┆ 1     │
        ╰───────┴───────┴───────┴───────╯
        <BLANKLINE>
        (Showing first 3 of 3 rows)
    """
    return Expression._call_builtin_scalar_fn("least", *exprs)


def is_nan(expr: Expression) -> Expression:
    """Checks if values are NaN (a special float value indicating not-a-number).

    Returns:
        Expression: Boolean Expression indicating whether values are invalid.

    Note:
        Nulls will be propagated! I.e. this operation will return a null for null values.

    Examples:
        >>> import daft
        >>> from daft.functions import is_nan
        >>>
        >>> df = daft.from_pydict({"data": [1.0, None, float("nan")]})
        >>> df = df.select(is_nan(df["data"]))
        >>> df.collect()
        ╭───────╮
        │ data  │
        │ ---   │
        │ Bool  │
        ╞═══════╡
        │ false │
        ├╌╌╌╌╌╌╌┤
        │ None  │
        ├╌╌╌╌╌╌╌┤
        │ true  │
        ╰───────╯
        <BLANKLINE>
        (Showing first 3 of 3 rows)

    """
    return Expression._call_builtin_scalar_fn("is_nan", expr)


def is_inf(expr: Expression) -> Expression:
    """Checks if values in the Expression are Infinity.

    Returns:
        Expression: Boolean Expression indicating whether values are Infinity.

    Note:
        Nulls will be propagated! I.e. this operation will return a null for null values.

    Examples:
        >>> import daft
        >>> from daft.functions import is_inf
        >>>
        >>> df = daft.from_pydict({"data": [-float("inf"), 0.0, float("inf"), None]})
        >>> df = df.select(is_inf(df["data"]))
        >>> df.collect()
        ╭───────╮
        │ data  │
        │ ---   │
        │ Bool  │
        ╞═══════╡
        │ true  │
        ├╌╌╌╌╌╌╌┤
        │ false │
        ├╌╌╌╌╌╌╌┤
        │ true  │
        ├╌╌╌╌╌╌╌┤
        │ None  │
        ╰───────╯
        <BLANKLINE>
        (Showing first 4 of 4 rows)

    """
    return Expression._call_builtin_scalar_fn("is_inf", expr)


def not_nan(expr: Expression) -> Expression:
    """Checks if values are not NaN (a special float value indicating not-a-number).

    Returns:
        Expression: Boolean Expression indicating whether values are not invalid.

    Note:
        Nulls will be propagated! I.e. this operation will return a null for null values.

    Examples:
        >>> import daft
        >>> from daft.functions import not_nan
        >>>
        >>> df = daft.from_pydict({"x": [1.0, None, float("nan")]})
        >>> df = df.select(not_nan(df["x"]))
        >>> df.collect()
        ╭───────╮
        │ x     │
        │ ---   │
        │ Bool  │
        ╞═══════╡
        │ true  │
        ├╌╌╌╌╌╌╌┤
        │ None  │
        ├╌╌╌╌╌╌╌┤
        │ false │
        ╰───────╯
        <BLANKLINE>
        (Showing first 3 of 3 rows)

    """
    return Expression._call_builtin_scalar_fn("not_nan", expr)


def fill_nan(expr: Expression, fill_value: Expression) -> Expression:
    """Fills NaN values in the Expression with the provided fill_value.

    Returns:
        Expression: Expression with Nan values filled with the provided fill_value

    Examples:
        >>> import daft
        >>> from daft.functions import fill_nan
        >>>
        >>> df = daft.from_pydict({"data": [1.1, float("nan"), 3.3]})
        >>> df = df.with_column("filled", fill_nan(df["data"], 2.2))
        >>> df.show()
        ╭─────────┬─────────╮
        │ data    ┆ filled  │
        │ ---     ┆ ---     │
        │ Float64 ┆ Float64 │
        ╞═════════╪═════════╡
        │ 1.1     ┆ 1.1     │
        ├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
        │ NaN     ┆ 2.2     │
        ├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
        │ 3.3     ┆ 3.3     │
        ╰─────────┴─────────╯
        <BLANKLINE>
        (Showing first 3 of 3 rows)

    """
    return Expression._call_builtin_scalar_fn("fill_nan", expr, fill_value)
